---
title: "R Notebook"
output: html_notebook
---

Install and Load DESeq2
```{r, echo = FALSE}
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
library(DESeq2)
```

Also load readxl 
```{r, echo = FALSE}
library('readxl')
```

Load the full expression matrix for 3D_soft and 3D_stiff substrates
```{r}
soft <- read_excel("3D_NF_so.xlsx")
stiff <- read_excel("3D_NF_st.xlsx")
```

Create a expression martrix for CAF samples grown on 3D_soft and 3D_stiff substrates
note: we want to keep the ENSid for each gene, so we keep the first column. 
```{r}
countData <- data.frame(soft[,c(1, 5, 6, 7)], stiff[, c(5, 6, 7)])
```

Now we need to create a matrix for the metadata associated to each sample. 
This will enable DESeq2 to recongnise which sample is associated to each condition.
This way we can compare the gene expression of CAF grown on 3D_soft substrate 
(the control case), with the gene expression of CAF gwon on 3D_stiff 
(the 'treatment' case). 
```{r}
control <- 'control'
treatment <- 'treament'
id <- c('soft1', 'soft2', 'soft3', 'stiff1', 'stiff2', 'stiff3')
dex <- c(rep(control, 3), rep(treatment, 3))

metadata <-  data.frame(id=as.factor(id), dex=as.factor(dex))
metadata
```

Using the metadata matrix and the expression matrix we can perform 
Differential Gene Expression analyisis using DESeq2
We will discard all the lines that have at least two cells with 9 or less counts, 
this will help us with the False Discovery Rate. 
Also we will use a pvalue of 0.05. 
```{r}
dds <- DESeqDataSetFromMatrix(countData=countData, 
                              colData=metadata, 
                              design= ~ dex, tidy = TRUE)
keep <- rowSums(counts(dds) >= 10) >= 2
dds <- dds[keep,]
dds <- DESeq(dds)
res <- results(dds,alpha = 0.05, pAdjustMethod = 'BH')
res
```

As you can see from the results we still have almost 30% of the genes with 
low counts and only 1 outliers. 

Interestling there are almosto 3000 genes that are differentialy expressed(de)! 

We use a Fold Change (FC) of at least |2| (absolute value) and 
a pvalue adjusted of less than 0.05 to state that a gene is 
de. This will reduce the number of de. 
We can store the results in a data frame and label the differentialy expressed 
genes with UP or DOWN if the gene is up regulated or down regualed, respectively.
Last but not least we can save the data frame in a excel file.
```{r}
de <- data.frame(res)
de$diffexpressed <- "NO"
de$diffexpressed[de$log2FoldChange > 2 & de$padj < 0.05] <- "UP"
de$diffexpressed[de$log2FoldChange < -2 & de$padj < 0.05] <- "DOWN"
```


---------------------------------------------------------------------------------------
                Sanity check and shrinking of the data sets
---------------------------------------------------------------------------------------

Load the TCGA-KIRC.htseq_fpkm data set. For each patient we have the expression 
level of almost all the genes. Also load the phenotype data set of the TCGA-KIRC 
project. 
```{r}
geno_data <- read.delim("~/Documents/R for Bs/KIRC pj/TGCA data/TCGA-KIRC.htseq_fpkm.tsv.gz.tsv")
pheno_data <- read.delim("~/Documents/R for Bs/KIRC pj/TGCA data/TCGA-KIRC.GDC_phenotype.tsv.gz.tsv")
```

Define the sequenced patients from the genomic data set
```{r}
sequenced_patients <- colnames(geno_data)
all <- intersect(gsub('-','.',pheno_data$submitter_id.samples),sequenced_patients)
all[seq(1,10)]
```
The two datasets contain information about the same 607 patients

```{r}
pheno_data <-pheno_data[which(gsub('-','.',pheno_data$submitter_id.samples) %in% all),]

pheno_data <- pheno_data[-(which(pheno_data$sample_type.samples != 'Primary Tumor')),]

#just to be sure
which(pheno_data$sample_type.samples != 'Primary Tumor')

which(pheno_data$tumor_stage.diagnoses=='not reported')
```

The corresponding submitter id are:
```{r}
submitter_not_reported <- pheno_data$submitter_id.samples[which(pheno_data$tumor_stage.diagnoses=='not reported')]
submitter_not_reported
```

Remove from the pheno data set the rows of these 4 patients
```{r}
pheno_data <- pheno_data[-c(115, 423, 504),]
```

Remove from the geno data set the columns of these 4 patients
```{r}
geno_data[,'TCGA.B4.5838.01A'] <- NULL 
geno_data[,'TCGA.MM.A563.01A'] <- NULL 
geno_data[,'TCGA.BP.4798.01A'] <- NULL 
```


```{r}
all <- intersect(gsub('-','.',pheno_data$submitter_id.samples),sequenced_patients)
```

Get also the Ensembl_id without the version number
```{r}
Ensembl_id <- geno_data[,1]
Ensembl_id <- gsub("\\.[0-9]*$", "", Ensembl_id)
```

Update the geno_data with only the patientid of tumoral samples and keeping the 
Ensembl_id
```{r}
geno_data <- data.frame(Ensembl_id, geno_data[,which(colnames(geno_data) %in% all)])
sequenced_patients <- colnames(geno_data)[-1]
```

---------------------------------------------------------------------------------------
                  Calculating the MeCo Score 
---------------------------------------------------------------------------------------

Lists of gene UP and DOWN regualted
```{r}
UP <- rownames(de[which(de$diffexpressed == 'UP'),])
DOWN <- rownames(de[which(de$diffexpressed == 'DOWN'),])
```

Creating two distinct expression matrix for genes UP or DOWN regulated. 
Both matrix are linked by the patientid. 
```{r}
geno_data_down <- geno_data[which(geno_data$Ensembl_id %in% DOWN), ]
geno_data_up <- geno_data[which(geno_data$Ensembl_id %in% UP), ]
```

Finally Calculating the MeCo :)
note: may be is better applay but i am too tired to try
```{r}
MecoScore <- data.frame(sequenced_patients) 
MecoScore$MeSc <- NA
for (i in 1:nrow(MecoScore)){
  MecoScore$MeSc[i] <- mean(as.numeric(geno_data_up[,i+1])) - mean(as.numeric(geno_data_down[,i+1]))
}

summary(MecoScore)
```



```{r}
#check presence of stage i patients in the genomic dataset
samples_stage1 <- pheno_data[which(pheno_data$tumor_stage.diagnoses=='stage i'),1]#take the first column containing the submitter id sample
samples_stage1 <- gsub('-','.', samples_stage1)#replace - with . as patients are reported differently in the two datasets
sequenced_stage1 <- intersect(sequenced_patients,samples_stage1)

#check presence of stage ii patients in the genomic dataset
samples_stage2 <- pheno_data[which(pheno_data$tumor_stage.diagnoses=='stage ii'),1]
samples_stage2 <- gsub('-','.', samples_stage2)
sequenced_stage2 <- intersect(sequenced_patients,samples_stage2)

#check presence of stage iii patients in the genomic dataset
samples_stage3 <- pheno_data[which(pheno_data$tumor_stage.diagnoses=='stage iii'),1]
samples_stage3 <- gsub('-','.', samples_stage3)
sequenced_stage3 <- intersect(sequenced_patients,samples_stage3)

#check presence of stage iv patients in the genomic dataset
samples_stage4 <- pheno_data[which(pheno_data$tumor_stage.diagnoses=='stage iv'),1]
samples_stage4 <- gsub('-','.', samples_stage4)
sequenced_stage4 <- intersect(sequenced_patients,samples_stage4)

mean(MecoScore[which(MecoScore$sequenced_patients %in% sequenced_stage1), 'MeSc'])
mean(MecoScore[which(MecoScore$sequenced_patients %in% sequenced_stage2), 'MeSc'])
mean(MecoScore[which(MecoScore$sequenced_patients %in% sequenced_stage3), 'MeSc'])
mean(MecoScore[which(MecoScore$sequenced_patients %in% sequenced_stage4), 'MeSc'])
```
---------------------------------------------------------------------------------------
                  Refining the Meco
---------------------------------------------------------------------------------------

Pathway analysis with Metascape 
```{r}
DifferentialExpressedGenes <- c(DOWN,UP)
longlistde <- data.frame(DifferentialExpressedGenes)
write_xlsx(LongListde, 'PathwayAnalysis2.xlsx')
```

Getting the excel sheet from Metascape
```{r}
metascape_results <- read_excel("~/Downloads/metascape_results2.xlsx")
metascape_results <- metascape_results[is.na(metascape_results$`Gene ID`) == FALSE,]
```

Grouping genes in distinct pathways
1) get all the disinct pathways

```{r}
distinct_pathways <- colnames(metascape_results)[-seq(1, 15)]
head(distinct_pathways)
```
2) For each pathway get the list of genes. List of Lists (LoL)
(this is a stupid way, but also the fastest!) 

To be improvedâ€¦
```{r}

sum(as.integer(metascape_results$`R-HSA-1474244 Extracellular matrix organizat`))

LoL <- list()

for(i in seq(1, length(distinct_pathways))){
  LoL[i] <- metascape_results[which(metascape_results[,(15 + i)] == '1.0'), 'DifferentialExpressedGenes']
}

names(LoL) <- distinct_pathways
```

Now we have to manually curate the main pathways (yeah! more fun!)
Looking at the top 20 pathways from Metascape we selected the 5 most promising path. 

```{r}
ECM_path <- unique(c(LoL$`M5884 NABA CORE MATRISOME`,LoL$`R-HSA-1474244 Extracellular matrix organizat`, LoL$`GO:0030198 extracellular matrix organizat`, LoL$`M5882 NABA PROTEOGLYCANS`, LoL$`R-HSA-1474228 Degradation of the extracellul`, LoL$`R-HSA-2129379 Molecules associated with elas`))

Proliferation_path <- unique(c(LoL$`GO:0001501 skeletal system development`, LoL$`GO:0061061 muscle structure development`))

Chemotaxis_path <- LoL$`GO:0006935 chemotaxis`

Inflammatory_path <- LoL$`GO:0006954 inflammatory response`

Antitumoral_path <- LoL$`GO:0008285 negative regulation of cell po`
```

Dived the genes into Up and Down regulated for each path
```{r}
ECM_path_Down <- intersect(ECM_path, DOWN)
ECM_path_Up <- intersect(ECM_path, UP)

Proliferation_path_Down <- intersect(Proliferation_path, DOWN)
Proliferation_path_Up <- intersect(Proliferation_path, UP)

Chemotaxis_path_Down <- intersect(Chemotaxis_path, DOWN)
Chemotaxis_path_Up <- intersect(Chemotaxis_path, UP)

Inflammatory_path_Down <- intersect(Inflammatory_path, DOWN)
Inflammatory_path_Up <- intersect(Inflammatory_path, UP)

Antitumoral_path_Down <- intersect(Antitumoral_path, DOWN)
Antitumoral_path_Up <- intersect(Antitumoral_path, UP)
```

Calculate for each path a specific MeCo score! 
This is a refined MeCo because it is pathwway based. 

MeCo for ECM
```{r}
geno_data_ECMdown <- geno_data[which(geno_data$Ensembl_id %in% ECM_path_Down), ]
geno_data_ECMup <- geno_data[which(geno_data$Ensembl_id %in% ECM_path_Up), ]

MecoScoreECM <- data.frame(sequenced_patients) 
MecoScoreECM$MeSc <- NA
for (i in 1:nrow(MecoScoreECM)){
  MecoScoreECM$MeSc[i] <- mean(as.numeric(geno_data_ECMup[,i+1])) - mean(as.numeric(geno_data_ECMdown[,i+1]))
}

summary(MecoScoreECM)
```

MeCo for Proliferation 
```{r}
geno_data_down <- geno_data[which(geno_data$Ensembl_id %in% Proliferation_path_Down), ]
geno_data_up <- geno_data[which(geno_data$Ensembl_id %in% Proliferation_path_Up), ]

MecoScorePro <- data.frame(sequenced_patients) 
MecoScorePro$MeSc <- NA
for (i in 1:nrow(MecoScorePro)){
  MecoScorePro$MeSc[i] <- mean(as.numeric(geno_data_up[,i+1])) - mean(as.numeric(geno_data_down[,i+1]))
}

summary(MecoScorePro)
```

MeCo for Chemotaxis
```{r}
geno_data_down <- geno_data[which(geno_data$Ensembl_id %in% Chemotaxis_path_Down), ]
geno_data_up <- geno_data[which(geno_data$Ensembl_id %in% Chemotaxis_path_Up), ]

MecoScoreCh <- data.frame(sequenced_patients) 
MecoScoreCh$MeSc <- NA
for (i in 1:nrow(MecoScoreCh)){
  MecoScoreCh$MeSc[i] <- mean(as.numeric(geno_data_up[,i+1])) - mean(as.numeric(geno_data_down[,i+1]))
}

summary(MecoScoreCh)
```

MeCo for Inflammation
```{r}
geno_data_down <- geno_data[which(geno_data$Ensembl_id %in% Inflammatory_path_Down), ]
geno_data_up <- geno_data[which(geno_data$Ensembl_id %in% Inflammatory_path_Up), ]

MecoScoreInf <- data.frame(sequenced_patients) 
MecoScoreInf$MeSc <- NA
for (i in 1:nrow(MecoScoreInf)){
  MecoScoreInf$MeSc[i] <- mean(as.numeric(geno_data_up[,i+1])) - mean(as.numeric(geno_data_down[,i+1]))
}

summary(MecoScoreInf)
```
MeCo for Antitumoral 

```{r}
geno_data_down <- geno_data[which(geno_data$Ensembl_id %in% Antitumoral_path_Down), ]
geno_data_up <- geno_data[which(geno_data$Ensembl_id %in% Antitumoral_path_Up), ]

MecoScoreAnt <- data.frame(sequenced_patients) 
MecoScoreAnt$MeSc <- NA
for (i in 1:nrow(MecoScoreAnt)){
  MecoScoreAnt$MeSc[i] <- mean(as.numeric(geno_data_up[,i+1])) - mean(as.numeric(geno_data_down[,i+1]))
}

summary(MecoScoreAnt)
```

---------------------------------------------------------------------------------------
                    Testing the fined MeCos  
---------------------------------------------------------------------------------------

MeCoECM in all 4 stages
```{r}
mean(MecoScoreECM[which(MecoScoreECM$sequenced_patients %in% sequenced_stage1), 'MeSc'])
mean(MecoScoreECM[which(MecoScoreECM$sequenced_patients %in% sequenced_stage2), 'MeSc'])
mean(MecoScoreECM[which(MecoScoreECM$sequenced_patients %in% sequenced_stage3), 'MeSc'])
mean(MecoScoreECM[which(MecoScoreECM$sequenced_patients %in% sequenced_stage4), 'MeSc'])
```

MeCoProliferation in all 4 stages
```{r}
mean(MecoScorePro[which(MecoScorePro$sequenced_patients %in% sequenced_stage1), 'MeSc'])
mean(MecoScorePro[which(MecoScorePro$sequenced_patients %in% sequenced_stage2), 'MeSc'])
mean(MecoScorePro[which(MecoScorePro$sequenced_patients %in% sequenced_stage3), 'MeSc'])
mean(MecoScorePro[which(MecoScorePro$sequenced_patients %in% sequenced_stage4), 'MeSc'])
```
MeCoChemotaxis in all 4 stages
```{r}
mean(MecoScoreCh[which(MecoScoreCh$sequenced_patients %in% sequenced_stage1), 'MeSc'])
mean(MecoScoreCh[which(MecoScoreCh$sequenced_patients %in% sequenced_stage2), 'MeSc'])
mean(MecoScoreCh[which(MecoScoreCh$sequenced_patients %in% sequenced_stage3), 'MeSc'])
mean(MecoScoreCh[which(MecoScoreCh$sequenced_patients %in% sequenced_stage4), 'MeSc'])
```
MeCoInflammation in all 4 stages
```{r}
mean(MecoScoreInf[which(MecoScoreInf$sequenced_patients %in% sequenced_stage1), 'MeSc'])
mean(MecoScoreInf[which(MecoScoreInf$sequenced_patients %in% sequenced_stage2), 'MeSc'])
mean(MecoScoreInf[which(MecoScoreInf$sequenced_patients %in% sequenced_stage3), 'MeSc'])
mean(MecoScoreInf[which(MecoScoreInf$sequenced_patients %in% sequenced_stage4), 'MeSc'])
```
MeCoAntitumoral in all 4 stages
```{r}
mean(MecoScoreAnt[which(MecoScoreAnt$sequenced_patients %in% sequenced_stage1), 'MeSc'])
mean(MecoScoreAnt[which(MecoScoreAnt$sequenced_patients %in% sequenced_stage2), 'MeSc'])
mean(MecoScoreAnt[which(MecoScoreAnt$sequenced_patients %in% sequenced_stage3), 'MeSc'])
mean(MecoScoreAnt[which(MecoScoreAnt$sequenced_patients %in% sequenced_stage4), 'MeSc'])
```
tap tap tap

tap tap tap








